src/OpenLoco/S5/SawyerStream.cpp

```diff
@@ -0,0 +1,488 @@
+            decodeRotate(_decodeBuffer2, data);
+            return _decodeBuffer2.getSpan();
+        default:
+            throw std::runtime_error(exceptionUnknownEncoding);
+    }
+}
+
+void SawyerStreamReader::decodeRunLengthSingle(FastBuffer& buffer, stdx::span<uint8_t const> data)
+{
+    for (size_t i = 0; i < data.size(); i++)
+    {
+        uint8_t rleCodeByte = data[i];
+        if (rleCodeByte & 128)
+        {
+            i++;
+            if (i >= data.size())
+            {
+                throw std::runtime_error(exceptionInvalidRLE);
+            }
+
+            auto copyLen = static_cast<size_t>(257 - rleCodeByte);
+            auto copyByte = data[i];
+            buffer.push_back(copyByte, copyLen);
+        }
+        else
+        {
+            if (i + 1 >= data.size() || i + 1 + rleCodeByte + 1 > data.size())
+            {
+                throw std::runtime_error(exceptionInvalidRLE);
+            }
+
+            auto copyLen = static_cast<size_t>(rleCodeByte + 1);
+            buffer.push_back(&data[i + 1], copyLen);
+            i += rleCodeByte + 1;
+        }
+    }
+}
+
+void SawyerStreamReader::decodeRunLengthMulti(FastBuffer& buffer, stdx::span<uint8_t const> data)
+{
+    for (size_t i = 0; i < data.size(); i++)
+    {
+        if (data[i] == 0xFF)
+        {
+            i++;
+            if (i >= data.size())
+            {
+                throw std::runtime_error(exceptionInvalidRLE);
+            }
+            buffer.push_back(data[i]);
+        }
+        else
+        {
+            auto offset = static_cast<int32_t>(data[i] >> 3) - 32;
+            assert(offset < 0);
+            if (static_cast<size_t>(-offset) > buffer.size())
+            {
+                throw std::runtime_error(exceptionInvalidRLE);
+            }
+            auto copySrc = buffer.data() + buffer.size() + offset;
+            auto copyLen = static_cast<size_t>((data[i] & 7) + 1);
+
+            // Copy it to temp buffer first as we can't copy buffer to itself due to potential
+            // realloc in between reserve and push
+            uint8_t copyBuffer[32];
+            assert(copyLen <= sizeof(copyBuffer));
+            std::memcpy(copyBuffer, copySrc, copyLen);
+            buffer.push_back(copyBuffer, copyLen);
+        }
+    }
+}
+
+void SawyerStreamReader::decodeRotate(FastBuffer& buffer, stdx::span<uint8_t const> data)
+{
+    uint8_t code = 1;
+    for (size_t i = 0; i < data.size(); i++)
+    {
+        buffer.push_back(ror(data[i], code));
+        code = (code + 2) & 7;
+    }
+}
+
+SawyerStreamWriter::SawyerStreamWriter(const fs::path& path)
+{
+    _stream.exceptions(std::ifstream::failbit);
+    _stream.open(path, std::ios::out | std::ios::binary);
+}
+
+void SawyerStreamWriter::writeChunk(SawyerEncoding chunkType, const void* data, size_t dataLen)
+{
+    auto encodedData = encode(chunkType, stdx::span<uint8_t const>(reinterpret_cast<const uint8_t*>(data), dataLen));
```

Comment:
I find this line a little inelegant due to the type being specified twice. Is there a way to avoid that?

Reply:
ok done